app/src/main/java/space/taran/arknavigator/mvp/model/repo/tags/TagsStorageRepo.kt

```diff
@@ -11,29 +13,31 @@ class TagsStorageRepo(
     private val foldersRepo: FoldersRepo,
     private val indexRepo: ResourcesIndexRepo
 ) {
+    private val provideMutex = Mutex()
     private val storageByRoot = mutableMapOf<Path, PlainTagsStorage>()
 
     suspend fun provide(rootAndFav: RootAndFav): TagsStorage = withContext(Dispatchers.IO) {
-        val allRoots = foldersRepo.query().succeeded.keys
-        val roots = if (rootAndFav.isAllRoots()) allRoots else listOf(rootAndFav.root!!)
+        provideMutex.withLock {
+            val roots = foldersRepo.resolveRoots(rootAndFav)
 
-        val storageShards = roots.map { root ->
-            val index = indexRepo.provide(root)
-            val resources = index.listAllIds()
-            if (storageByRoot[root] != null) {
-                val storage = storageByRoot[root]!!
-                storage.checkResources(resources)
-                storage
-            } else {
-                val fresh = PlainTagsStorage(root, resources)
-                fresh.init()
-                fresh.cleanup(resources)
-                storageByRoot[root] = fresh
-                fresh
+            val storageShards = roots.map { root ->
+                val index = indexRepo.provide(root)
+                val resources = index.listAllIds()
+                if (storageByRoot[root] != null) {
```

Comment:
Similar to `ResourcesIndexRepo.provide()`, "read-write lock" would allow us to really lock only when we actually write something into the map.